### DISCUSSION TITLE
阅读《Nacos 架构 & 原理》有感

### DISCUSSION THREAD

**Discussion by zhicheng-ning** [X-lab2017/open-wonderland#48] [4️⃣ 新媒体]
# 收获
- 注重自研、开源、商业化三位一体，讲清开源的价值，否则开源除了情怀和展示技术影响力之外很难对一个商业公司产生实实在在的价值
- 在整个行业的主流技术完全定型之前，什么时候开源都为时不晚
- 注重客户的使用需求，不是所有公司都有足够大的体量来支持应用的多模块拆分
- 以开源为内核，以商业化为扩展；开源做生态，商业化做企业级特性，阿里内部做性能和高可用；开源做组件，商业化做解决方案

# 1. 首先是谈到阿里为什么会做开源？
阿里做开源大概有两个阶段。
- 第一个阶段是 2018 年之前，取之于开源，反哺于社区，开源是一种情怀，是一种文化，是一种展示技术影响力和技术实力的方式。阿里凭借着互联网场景和规模的优势走在了时代的前列，完成了去 IOE ，创造了企业级互联网架构等壮举，并且开源了很多自主产品如 Dubbo、RocketMQ、Tengine、Jstorm 等，产生了巨大的影响力，在互联网行业广泛使用，**但是这一阶段的开源除了情怀和展示技术影响力之后很难量化对公司的价值，因此也比较难以持续发展**。
- 第二个阶段是 2018 年开始，随着云计算发展，开源作为一种标准加速云计算发展，尤其 K8s 迅速崛起给了阿里很大的启示，因此阿里巴巴也在 2018 年制定了一个全面开源，加速企业数字化转型的战略目标。这个阶段的开源发生了本质的两个变化。**第一更重视社区和生态建设，第二更注重自研、开源、商业化三位一体，讲清开源的价值，能够持续投入开源，解决第一阶段难以持续的问题**。

# 2. 2018 年开源是否晚了？如何定位和打造竞争力？
- 相比当时比较流行的竞品（Netflix 公司开源的 Eureka），Nacos 此时开源确实晚了一些，但是相比于整个行业其实不晚，因为当时云原生和微服务整个普及度还很低。当时为什么开源还有两个原因：
- 第一个点是 Nacos 当时是一个闭源的一个软件，经常有业务方跳出来说你看 Eureka 多好，你们哪里哪里不行，如果阿里不开源出去打一打，怎么更好的证明 Nacos 更好
- 有二个点是当时阿里是有商业化产品的，**虽然我们知道我们更好，但是奈何用户选择的是 Eureka，我们只能兼容，而且我们不出去，不成为默认标准，不知道未来还要被迫兼容更多不如我们的产品**，这对我们来说是一个灾难

# 3. 内部有三个产品，是开源三个产品还是合成一个产品开源 ？
由于当时 Spring-cloud 的崛起，微服务多个模块逐步被划分，包括注册中心、配置中心，**如果从产品定位上，期望定位简单清晰，利于传播，我们需要分别开源我们内部产品**，这样又会分散我们品牌和运营资源。另外**大部分客户没有阿里这么大的体量，模块拆分过细，部署和运维成本都会成倍上涨**，而且阿里巴巴也是从最早一个产品逐步演化成 3 个产品的，因此我们最终决定将内部三个产品合并统一开源

# 4. 开源产品跟商业化产品的关系是什么，是否会削弱商业化产品的竞争力？
随着我们选择三合一的开源模式，又面临另外一个问题，未来内部和商业化关系是什么，代码关系是什么？
这个问题应该说一直持续，但是我们定下来**开源、自研、商业化三位一体**的战略，**以开源为内核，以商业化为扩展；开源做生态，商业化做企业级特性，阿里内部做性能和高可用；开源做组件，商业化做解决方案**；并且随着时间推移，基本按照这思路完成的正循环，全面系统的打造了 Nacos 各个维度的能力。

**Comment by will-ww** [comment]
写得好！👍

Nacos 也是阿里商业化产品背后的重要力量，特别是有阿里云这个业务，因此就能不断获得公司资源发展开源项目，是一个挺好的成功案例。开源的可持续性确实非常需要商业的支持~

ps. 书籍的相关链接等信息，是否可以补充下~ @zhicheng-ning

**Reply by zhicheng-ning** [reply]
> 书籍的相关链接等信息

好的~，https://www.yuque.com/nacos/ebook

**Reply by will-ww** [reply]
加一个知识库的版本：https://xlab2017.yuque.com/msdpvs/fzem1h/26208088
